addons/misra.py

```diff
@@ -1495,6 +1495,248 @@ def misra_8_14(self, rawTokens):
+
+        # Checks if the initializer conforms to the elements of the struct or union
+        # Parameters:
+        #   token:      root node of the initializer tree
+        #   elements:   the elements as specified in the declaration
+        def checkObjectInitializer(token, elements):
+            if not token:
+                return True
+
+            # Initializer must start with a curly bracket
+            if not token.str == '{':
+                self.reportError(token, 9, 2)
+                return False
+
+            # Empty initializer is not ok { }
+            if not token.astOperand1:
+                self.reportError(token, 9, 2)
+                return False
+
+            token = token.astOperand1
+
+            # Zero initializer is ok { 0 }
+            if token.str == '0' :
+                return True
+
+            pos = None
+            while(token):
+                if token.str == ',':
+                    token = token.astOperand1
+                else:
+                    if pos == None:
+                        pos = 0
+
+                    if token.isAssignmentOp:
+                        if token.astOperand1.str == '.':
+                            elementName = token.astOperand1.astOperand1.str
+                            pos = next((i for i, element in enumerate(elements) if element.name == elementName), len(elements))
+                        token = token.astOperand2
+
+                    if pos >= len(elements):
+                        self.reportError(token, 9, 2)
+                        return False
+
+                    element = elements[pos]
+                    if element.elementType == 'class':
+                        if token.isName:
+                            if not token.valueType.typeScope  == element.valueType.typeScope:
+                                self.reportError(token, 9, 2)
+                                return False
+                        else:
+                            subElements = getRecordElements(element.valueType)
+                            if not checkObjectInitializer(token, subElements):
+                                return False
+                    elif element.elementType == 'array':
+                        if not checkArrayInitializer(token, element.dimensions, element.valueType):
+                            return False
+                    elif token.str == '{':
+                        self.reportError(token, 9, 2)
+                        return False
+
+                    # The assignment represents the astOperand
+                    if token.astParent.isAssignmentOp:
+                        token = token.astParent
+
+                    if not token == token.astParent.astOperand2:
+                        pos = pos + 1
+                        token = token.astParent.astOperand2
+                    else:
+                        token = None
+
+            return True
+
+        # ------
+        for token in data.tokenlist:
```

Comment:
You can loop through each variable instead..

    for var in data.variables:
        ... check if initialization is proper ...